### ISSUE TITLE
Refactor SwiftTask to use Operations instead.

### ISSUE THREAD

**Issue by danthorpe** [tinode/swift#5]
Having previously gone down a WebSocket + FRP route, having thought about all the difficulties, I think it's better to use Operations. For these reasons...

- [ ] Ability to attach an `OperationCondition` which in turn will create the web socket connection.
- [ ] Base `Operation` subclass to write message to the socket.
- [ ] `RepeatedOperation` to listen to socket for receiving messages

For this to work we'll need a class which will "own" the connection - and then ideally a way to get the only possible instance. Not necessarily a singleton, but more likely something like `networkForEnvironment() -> Network` or something.

**Comment by or-else** [comment]
Hey. I don't know much about Swift or iOS programming so I can't comment specifically. In general both JS and Android have Connection isolated into a class which takes callbacks as well as returns a promise resolved on connection completion. Messages and dropped connection are reported through callbacks.

**Comment by danthorpe** [comment]
@or-else sure, FRP is one way to go, and I've done it on a project in Swift before with WebSocket & Thrift. Ultimately felt it led to hard to maintain code which lacked cohesion, and that was with the benefit of strong typing thanks to Thrift.

I've been thinking about this for a bit and while the web socket under the hood will/does have an architecture like you're suggesting, the client/server which sits on top of it could benefit from things like dependencies, conditions & observers, and I think it'll just be much more modular to use discrete classes for things like sending messages.

**Comment by or-else** [comment]
I don't think I'm qualified to comment on your suggestions. You know better.

JFYI, both JS & Android clients define two main objects - Tinode and Topic. Tinode exposes lower-level API, maintains the websocket connection, client-side state/cache, routes messages to topics. The user is expected to communicate with the server primarily through the Topic objects. Here are the JS docs: http://tinode.github.io/js-api/

**Comment by danthorpe** [comment]
Okay - thanks - that's helpful, will review it this week.

At the moment, I guess I'm mostly working on the low level stuff. Getting the `WebSocket` type sorted out. I'm mostly using this: [SwiftWebSocket](https://github.com/tidwall/SwiftWebSocket), although it isn't really designed to support effective testing. So, I'm making a few changes.

Then a `Network` type, or similar, will own an instance of the socket and support doing things like connect/disconnect/send/receive etc - but this will be done using operations & queues. See [Operations](https://github.com/danthorpe/Operations).

On top of this, would sit something which is what the framework consumer works with - I haven't fully figured this bit out yet, so will look through how you're doing it in JS.

Most of this stuff (above) is general and will likely just be a dependency in the Tinode Swift stack. I have another repo, currently private which I'm setting some WebSocket stuff up in, should open it up in a week or so once it's a bit more complete.

**Comment by or-else** [comment]
If I understand correctly, Operation is a form of Promise (https://en.wikipedia.org/wiki/Futures_and_promises). Which makes sense. Tinode's both JS & Android use Promises a lot. The abstraction is pierced though. First, receiving messages requires a callback because messages arrive unsolicited without any actions on the client side. Second, one client request may generate multiple messages from the server. A promise is resolved on the first messages, the other messages require callbacks.

**Comment by danthorpe** [comment]
> If I understand correctly, Operation is a form of Promise 

Sort of, not exactly. Essentially, `Operation` is a `NSOperation` subclass which just performs *work* on a queue with the concept of dependencies. And `Operation` improves the state machine around this to provide well defined finishing states. This in turns makes it possible to support observers. Which allows us to do some work, and then do more work using the results of the previous operation. i.e. a Promise.

Using Promises or Functional Reactive Programming (FRP) (e.g. [PromiseKit](http://promisekit.org), or [ReactiveCocoa](http://reactivecocoa.io)) for this has a number of drawbacks:

These frameworks encourage functional programming, and "pure" functions, maintaining state is relatively alien. The goal with these techniques is to transform state. And, in general, I'm a proponent of functional programming and removal of state. However, in this case, we've got a websocket - which is inherently very statefull.

Improved Code-reuse: Given the above - there is a bunch of stuff which needs to happen before we can send a message (or do anything). E.g. create a web socket, open the web socket, register callbacks etc. These are all dependent tasks. Using promises, it is therefore necessary to write it like this (in pseudo code)

    ```
    createWebSocket()
        .then(open)
        .then(registerCallbacks)
        .then(sendMessage)
        .then(readMessage)
    ```

    In other words it must be fairly well defined in terms of the order in which things happen. Versus doing this using Operations, which is more like:

    ```
    let send = SendMessage(websocket)
    send.addCondition(OpenWebSocket(websocket))
    send.addCondition(RegisterCallbacks(websocket))
    send.addObserver(MessageReceived())
    queue.addOperation(send)
    ```

    The latter approach is an architecture which is readable, maintainable and testable. It also naturally leads to better code-reuse thanks to more discrete units, yet also, I think, gives overall better cohesion. I've come to this conclusion after writing an entire group messaging social networking app using the first approach, so I know how this works out when you're writing production code versus example pseudo code.

Now, I haven't exactly worked all how all the callbacks will work. But I do want to avoid the situation where the promise/operation is resolved/finished before all of the callbacks are received. But my vague idea is that sending a message will be a `GroupOperation`, which has the ability to keep itself running by adding child operations to its own internal queue. And, a "callback" will actually be another `Operation` (i.e. a child) which is waiting for something to come back on the socket. Therefore, `SendMessage` from the framework consumer's point of view, is actually a bunch of operations such as `WriteMessage` and one more `RegisterCallback` operations. This will allow us to use the same mechanisms to handle unsolicited messages from the server to the client, and messages from the client to the server which triggers one or more solicited messages back.

Obviously.. it's an evolving architecture.

**Comment by or-else** [comment]
I trust your opinion, I just don't get it (yet). The `a.then(b)` looks the same to me as `a.addCondition(b)` :) In the first case the promises are chained in a linked list, in the second there is an internal queue of operations. As for the order of execution being well defined, shouldn't it be well defined?

The only material difference is `queue.addOperation`, which seems to be the equivalent of `Executor` & `FutureTask` in Java. I did not use them on Android because the WS library had a non-blocking send().

**Comment by danthorpe** [comment]
Yeah, I'm not trying to suggest that this is a huge departure or wildly different, but just that it allows for a little bit more modularity and reuse.

For example, in the pseudo code above, I've added the two conditions to the operation after it's been initialized, but in reality they'd be added as part of the `SendMessage` class's init method, or actually more likely a base abstract NetworkMessage class, or something more generic.

With the promises, and purely functional approach I've found this a little trickier as you've always got to be accessing some `Promise<WebSocket, Error>`. Just figuring out how to get this when the app starts can potentially be or introduce a race condition, and we won't have any control over how the consumer kicks things off in their app (which in iOS terms is often a total mess depending on how experienced the first developer on the project was).

Granted, with an Operations based approach we're still going to need to get the single `WebSocket` instance which but this is pretty trivial.

Anyway, once I get a bit further along you'll soon see how it'll work out. Trust me! :)
